Add Quirks for UA on iPad for tiktok.com (macOS version)

brentfulgham opened the pull request:
#### da78f7a0a223588602c4a99abb339ad9deaf48b4
<pre>
Add Quirks for UA on iPad for tiktok.com (macOS version)
<a href="https://bugs.webkit.org/show_bug.cgi?id=291113">https://bugs.webkit.org/show_bug.cgi?id=291113</a>
&lt;<a href="https://rdar.apple.com/148467973">rdar://148467973</a>&gt;

Reviewed by Wenson Hsieh.

This patch builds on the work in Bug 290372 to ensure the same behavior on macOS
(as well as other WebKit ports).

It also adjust the flow in the iOS case to allow the quirk to be turned off if the
user deactivated custom website quirks.

* Source/WebKit/UIProcess/API/APIWebsitePolicies.h: Add a Move operator for setCustomUserAgent
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsCustomUserAgentOverride): Do more of the work in this method to
avoid duplicating it for macOS..
* Source/WebCore/page/Quirks.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp: Adjust compile macro now that both iOS and
macOS have an implementation of this.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies): Honor the &apos;Enable
Website Quirks&apos; setting and adopt the Move overload.
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies): Add logic for the
Quirk on macOS and adopt the Move overload.

Canonical link: <a href="https://commits.webkit.org/293441@main">https://commits.webkit.org/293441@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/31001b77c514fca11f4a597d967a402dfd3afd3d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8807 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104069 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49518 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100972 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18859 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27014 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75319 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32440 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101931 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14327 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89338 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55680 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14119 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7312 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48896 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84055 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7386 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106422 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26024 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84272 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26399 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85535 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83773 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21238 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28425 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6099 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19746 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25977 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31163 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25797 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29117 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27371 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14772,8 +14773,18 @@ void WebPageProxy::speechSynthesisResume(CompletionHandler<void()>&& completionH
+            // needsCustomUserAgentOverride() is currently very generic on purpose.
+            // In the future we want to pass more parameters for targeting specific domains.
+            if (Quirks::needsCustomUserAgentOverride(request.url()))
+                policies.setCustomUserAgent(makeStringByReplacingAll(standardUserAgentWithApplicationName(m_applicationNameForUserAgent), "like Gecko"_s, "like Gecko, like Chrome/136."_s));
```
What is happening when someone change the User agent in the Web Inspector for this site?

karlcow replied:
AH wait let me change my approved-non-approved-reviewer review ( :p )

```
In file included from /app/webkit/WebKitBuild/GTK/Release/DerivedSources/WebKit/unified-sources/UnifiedSource-88d1702b-9.cpp:2:
/app/webkit/Source/WebKit/UIProcess/WebPageProxy.cpp: In member function ‘WebKit::WebContentMode WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies(API::WebsitePolicies&, const WebCore::ResourceRequest&)’:
/app/webkit/Source/WebKit/UIProcess/WebPageProxy.cpp:14784:70: error: ‘standardUserAgentWithApplicationName’ was not declared in this scope
```

It breaks the build for WebKit GTK, obviously because it goes ino the way of their own UA override I guess, so there is a need to make extra-sur the code change makes it possible. 

1. benefit from it
2. do not break the current strategy they have use. 


which is at https://github.com/WebKit/WebKit/blob/9e0e1df49dfbdb36934d38e05319e4b85b9761b0/Source/WebCore/platform/glib/UserAgentGLib.cpp

karlcow reviewed (dismissed):
If it's working as intended (see my question), that would be a nice way to transfer at least temporarily the User Agent override from Safari to WebKit. 

LGTM (Though I'm not a reviewer so you will need someone else too.) modulo my questions about Web Inspector.

whsieh commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -973,6 +975,22 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+            // needsCustomUserAgentOverride() is currently very generic on purpose.
+            // In the future we want to pass more parameters for targeting specific domains.
+            if (auto customUserAgentForQuirk = Quirks::needsCustomUserAgentOverride(request.url(), m_applicationNameForUserAgent))
+                policies.setCustomUserAgent(*customUserAgentForQuirk);
```
```suggestion
                policies.setCustomUserAgent(WTFMove(*customUserAgentForQuirk));
```

brentfulgham replied:
Oh, good idea!

whsieh replied:
(Totally fine to do this as a followup if we need to change a bunch of other methods downstream to take && as well)

karlcow reviewed (approved):
looks good now!

karlcow commented on Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
```diff
@@ -973,6 +975,22 @@ static inline bool expectsLegacyImplicitRubberBandControl()
+    Ref preferences = m_preferences;
+    if (preferences->needsSiteSpecificQuirks()) {
+        if (policies.customUserAgent().isEmpty() && customUserAgent().isEmpty()) {
+            // FIXME (263619): This is done here for adding a UA override to tiktok. Should be in a common location.
+            // needsCustomUserAgentOverride() is currently very generic on purpose.
+            // In the future we want to pass more parameters for targeting specific domains.
```
@brentfulgham Maybe this comment is not needed anymore, now that the core of the method has been moved to Quirks as it should be.

brentfulgham committed:
Add Quirks for UA on iPad for tiktok.com (macOS version)
https://bugs.webkit.org/show_bug.cgi?id=291113
<rdar://148467973>

Reviewed by Wenson Hsieh.

This patch builds on the work in Bug 290372 to ensure the same behavior on macOS
(as well as other WebKit ports).

It also adjust the flow in the iOS case to allow the quirk to be turned off if the
user deactivated custom website quirks.

* Source/WebKit/UIProcess/API/APIWebsitePolicies.h: Add a Move operator for setCustomUserAgent
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::needsCustomUserAgentOverride): Do more of the work in this method to
avoid duplicating it for macOS..
* Source/WebCore/page/Quirks.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp: Adjust compile macro now that both iOS and
macOS have an implementation of this.
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies): Honor the 'Enable
Website Quirks' setting and adopt the Move overload.
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies): Add logic for the
Quirk on macOS and adopt the Move overload.

Canonical link: https://commits.webkit.org/293441@main
